Add TypeScript type definitions

Charrondev opened the pull request:
Fulfilling #269 This PR adds TypeScript type definitions to this project. These definitions build on top of the existing `@types/react` and `@types/quill` packages.

I've added definitions for:
- Mixin
- Component
- Toolbar

The dependancy versions of those are set to `*` so that they'll match whatever version that the user of this project has installed.

---

Closes #269

alexkrolick commented:
Excellent! I'd appreciate if someone with knowledge of Typescript could add a review.

Charrondev commented:
Maybe try adding the Hactoberfest label on this PR. That's how I found the initial issue.

zenoamaro commented on types.d.ts:
```diff
@@ -0,0 +1,76 @@
+    id: string;
+    className: string;
+    style: Object;
+    items: any[];
```
I think this is meant to be `QuillToolbarItems[]` instead of `any[]`.

zenoamaro commented on types.d.ts:
```diff
@@ -0,0 +1,76 @@
+    getEditor(): Quill.Quill;
+}
+
+export interface QuillToolbarItems {}
```
~~~ts
export interface QuillToolbarItems {
  type: string,
  label?: string,
  value?: string.
  items?: QuillToolbarItems[],
};
~~~

zenoamaro replied:
Also, this should be singular, ie. `QuillToolbarItem`.

zenoamaro commented on types.d.ts:
```diff
@@ -0,0 +1,76 @@
+    };
+}
+
+export interface QuillModules {
```
Let's just name these `Modules` to be consistent with `Toolbar*`.

zenoamaro commented:
Thanks a lot for the contribution, @Charrondev!

Typescript type definitions are a great addition.

Charrondev commented:
I finally got a chance to sit down and implement your feedback on this.
- Removed "Quill" prefixes from all of the exported elements
- Made all props optional
- Uncommented deprecated props

I've also added some comments on the deprecated properties with the stuff you have in the prop-types validation. In VSCode it looks like this if you try to use one of these deprecated properties:

![image](https://user-images.githubusercontent.com/1770056/32084508-9011a140-ba97-11e7-908a-8a4a9cd78975.png)

juanparadox commented on types.d.ts:
```diff
@@ -0,0 +1,76 @@
+    toolbar: string | ToolbarOptionObject | ToolbarOptionItem[];
+}
+
+export interface UnprivilegedEditor {
```
@Charrondev need help getting this merged by adding getContents() I can make the quick change if you don't have bandwidth to do so.

alexkrolick commented:
@zenoamaro any further comments or shall we go ahead and add this?

ajaska commented:
Looks like these changes went live about an hour ago and now I'm getting some errors

```
> tsc --project tsconfig.json --pretty --noEmit --noImplicitAny --noImplicitThis --alwaysStrict --strictNullChecks


29     getContents(index?: number, length?: number): DeltaStatic;}
                                                     ~~~~~~~~~~~

node_modules/react-quill/types.d.ts(29,51): error TS2304: Cannot find name 'DeltaStatic'.


100  setEditorReadOnly(editor: Quill.Quill, value: boolean);
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/react-quill/types.d.ts(100,2): error TS7010: 'setEditorReadOnly', which lacks return-type annotation, implicitly has an 'any' return type.


101  setEditorContents(editor: Quill.Quill, value: Quill.Delta | string);
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/react-quill/types.d.ts(101,2): error TS7010: 'setEditorContents', which lacks return-type annotation, implicitly has an 'any' return type.


102  setEditorSelection(editor: Quill.Quill, range: Quill.RangeStatic);
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/react-quill/types.d.ts(102,2): error TS7010: 'setEditorSelection', which lacks return-type annotation, implicitly has an 'any' return type.
```

ajaska commented:
Submitted a fix with #296
